sql/core/src/main/scala/org/apache/spark/sql/util/QueryExecutionListener.scala

```diff
@@ -81,7 +83,10 @@ class ExecutionListenerManager private[sql](
     loadExtensions: Boolean)
   extends Logging {
 
-  private val listenerBus = new ExecutionListenerBus(this, session)
+  // SPARK-39864: lazily create the listener bus on the first register() call in order to
+  // avoid listener overheads when QueryExecutionListeners aren't used:
+  @GuardedBy("this")
```

Comment:
I don't think there's a deadlock risk here (something that I always want to check whenever I'm adding synchronization to existing code).  

The register(), unregister(), clear(), and listListeners() methods call `ListenerBus` methods and those methods are synchronized. However, I don't think they'll have circular waits because those methods are just updating ListenerBus data structures and aren't running code which could call back into ExecutionListenerManager.

I'm a bit less sure about clone(), though.

To play it safe, I'm going to see if I can remove the method-level `synchronized` and instead do something more surgical.